pallets/oracle/src/lib.rs

```diff
@@ -0,0 +1,442 @@
+			self.public.clone()
+		}
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	/// Struct used to hold price data.
+	#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, scale_info::TypeInfo)]
+	pub struct PricePayload<Public, BlockNumber> {
+		block_number: BlockNumber,
+		price: u32,
+		public: Public
+	}
+
+	/// A bounded sized vector of recently submitted prices.
+	/// This is used to calculate average price.
+	#[pallet::storage]
+	#[pallet::getter(fn prices)]
+	pub(super) type Prices<T: Config> = StorageValue<_, BoundedVec<u32, T::MaxPrices>, ValueQuery>;
+
+	/// **StorageValue** that equals the **List of Clamor Account IDs** that both ***validate*** and ***send*** **unsigned transactions with signed payload**
+	///
+	/// NOTE: Only the Root User of the Clamor Blockchain (i.e the local node itself) can edit this list
+	#[pallet::storage]
+	pub type FragKeys<T: Config> = StorageValue<_, BTreeSet<ed25519::Public>, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn offchain_worker(block_number: T::BlockNumber) {
+
+			let res = Self::fetch_price_from_oracle(block_number);
+			if let Err(e) = res {
+				log::error!("Error: {}", e);
+			}
+		}
+	}
+
+	/// A public part of the pallet.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+
+		/// Add `public` to the **list of Clamor Account IDs** that can ***validate*** and ***send*** **unsigned transactions with signed payload**
+		///
+		/// NOTE: Only the Root User of the Clamor Blockchain (i.e the local node itself) can edit this list
+		#[pallet::weight(10000)]
+		pub fn add_key(origin: OriginFor<T>, public: ed25519::Public) -> DispatchResult {
+			ensure_root(origin)?;
+
+			log::debug!("New key: {:?}", public);
+
+			<FragKeys<T>>::mutate(|validators| {
+				validators.insert(public);
+			});
+
+			Ok(())
+		}
+
+		/// Remove a Clamor Account ID from `FragKeys`
+
+		/// NOTE: Only the Root User of the Clamor Blockchain (i.e the local node itself) can call this function
+		#[pallet::weight(10000)] // TODO
+		pub fn del_key(origin: OriginFor<T>, public: ed25519::Public) -> DispatchResult {
+			ensure_root(origin)?;
+
+			log::debug!("Removed key: {:?}", public);
+
+			<FragKeys<T>>::mutate(|validators| {
+				validators.remove(&public);
+			});
+
+			Ok(())
+		}
+
+		/// This function stores the new price received from the Oracle into a structure that contains a list of the last N-prices.
+		#[pallet::weight(10000)] // TODO
+		pub fn store_price(
+			origin: OriginFor<T>,
+			price_payload: PricePayload<T::Public, T::BlockNumber>,
+			_signature: T::Signature,
+		) -> DispatchResultWithPostInfo {
+			// This ensures that the function can only be called via unsigned transaction.
+			ensure_none(origin)?;
+
+			// TODO: regenerate the same Ethereum message received
+
+			Self::add_price(price_payload.block_number, price_payload.price);
```

Comment:
why we are adding? can you explain the logic? so to average multiple node prices? isn't that still dangerous? (One could have a x100000 pric)

Reply:
The idea here would be to have a set of last prices from where to calculate the average. This to avoid "oracle attacks" where someone could suddenly make the price of FRAG increase on an exchange, and use it to immediately buy a huge amount of NOVA. Having a bounded vector containing the last N prices from the oracle can avoid problems with sudden FRAG volatility.

Reply:
Where is this from? some paper? chainlink etc? or you invented it?

Reply:
This enormously simplify things here.